In [1]:
# 导入必要的库
import pandas as pd
import numpy as np


In [2]:
# 读取会员信息查询数据
member_df = pd.read_excel('会员信息查询.xlsx')

# 查看数据的基本信息
member_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 952714 entries, 0 to 952713
Data columns (total 12 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   会员卡号    952714 non-null  object        
 1   会员等级    952714 non-null  object        
 2   会员来源    952714 non-null  object        
 3   注册时间    952714 non-null  datetime64[ns]
 4   所属店铺编码  952714 non-null  object        
 5   门店店员编码  253828 non-null  object        
 6   省份      264801 non-null  object        
 7   城市      264758 non-null  object        
 8   性别      952714 non-null  object        
 9   生日      785590 non-null  object        
 10  年齡      952705 non-null  float64       
 11  生命级别    952714 non-null  object        
dtypes: datetime64[ns](1), float64(1), object(10)
memory usage: 87.2+ MB


有些列存在缺失值：
门店店员编码：只有253,828个非空值（约73%的数据缺失）
省份：只有264,801个非空值（约72%的数据缺失）
城市：只有264,758个非空值（约72%的数据缺失）
生日：有785,590个非空值（约18%的数据缺失）
年龄：有952,705个非空值（几乎没有缺失）

针对缺失严重的列直接删除，没有分析价值 省份、城市、门店店员编码

In [4]:
# 删除缺失严重的列
columns_to_drop = ['门店店员编码', '省份', '城市']
member_df = member_df.drop(columns=columns_to_drop)

# 查看处理后的数据结构
print("删除列后的数据结构:")
member_df.info()

删除列后的数据结构:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 952714 entries, 0 to 952713
Data columns (total 9 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   会员卡号    952714 non-null  object        
 1   会员等级    952714 non-null  object        
 2   会员来源    952714 non-null  object        
 3   注册时间    952714 non-null  datetime64[ns]
 4   所属店铺编码  952714 non-null  object        
 5   性别      952714 non-null  object        
 6   生日      785590 non-null  object        
 7   年齡      952705 non-null  float64       
 8   生命级别    952714 non-null  object        
dtypes: datetime64[ns](1), float64(1), object(7)
memory usage: 65.4+ MB


In [6]:
# 查看数据的前几行
print("数据前5行:")
print(member_df.head())

# 查看数值列的基本统计信息
print("\n数值列统计信息:")
print(member_df.describe())

# 查看分类列的分布情况
print("\n会员等级分布:")
print(member_df['会员等级'].value_counts())

print("\n会员来源分布:")
print(member_df['会员来源'].value_counts())

print("\n性别分布:")
print(member_df['性别'].value_counts())

print("\n生命级别分布:")
print(member_df['生命级别'].value_counts())

# 检查生日列的格式
print("\n生日列示例:")
print(member_df['生日'].head(10))

# 检查年龄的缺失值和异常值
print("\n年龄缺失值数量:", member_df['年齡'].isnull().sum())
print("年龄异常值(小于0或大于120):", 
      ((member_df['年齡'] < 0) | (member_df['年齡'] > 120)).sum())

数据前5行:
                    会员卡号  会员等级  会员来源                    注册时间  所属店铺编码 性别  \
0  BL6099033963550303769  黄金会员  线下扫码 2019-03-31 23:55:03.977  DPXX07  女   
1  BL6099033963450303763  黄金会员  线下扫码 2019-03-31 23:45:03.005  DPXX07  女   
2  BL6099033963464003767  白银会员  电商入口 2019-03-31 23:42:40.073  DPS00X  女   
3  BL6099033963460503766  黄金会员  线下扫码 2019-03-31 23:42:05.516  DPXX07  女   
4  BL6099033963660603765  白银会员  电商入口 2019-03-31 23:26:02.402  DPS00X  女   

           生日    年齡 生命级别  
0  1975-04-04  43.0   活跃  
1  1982-04-02  36.0   活跃  
2  1988-08-13  30.0   沉睡  
3  1993-11-24  25.0   活跃  
4  1993-03-20  26.0   沉睡  

数值列统计信息:
                                注册时间             年齡
count                         952714  952705.000000
mean   2018-03-04 14:43:11.089170176      28.968969
min              2017-08-03 00:00:00      -1.000000
25%              2017-08-03 00:00:00      24.000000
50%    2018-03-25 17:33:58.086499840      31.000000
75%    2018-09-01 13:21:11.283750144      39.000000
max   

数据概览：
会员等级：
大多数会员是白银会员(85%)，其次是黄金会员(15%)
铂金和钻石会员较少，分别只有1,123和489人
会员来源：
92%的会员来自线下扫码
8%的会员来自电商入口
性别分布：
约91%是女性会员
9%是男性会员
只有2个未知性别
生命级别：
84%的会员处于"沉睡"状态
约7%的会员"活跃"
约7%的会员"休眠"
约2%的会员处于"互动"状态
年龄：
平均年龄约29岁
有9个缺失值
有8个异常值（年龄<0或>120）
年龄范围从-1到147岁，明显有问题的数据
生日：
约17%的生日数据缺失
生日格式看起来正常，是字符串格式的日期

先处理年龄的异常值 然后转换生日数值类型 然后检车是否有重复会员

In [7]:
# 1. 处理年龄异常值
# 只将极端异常的年龄值设为NaN (小于0或大于120岁)
member_df.loc[(member_df['年齡'] < 0) | (member_df['年齡'] > 120), '年齡'] = np.nan
print("处理后的年龄缺失值数量:", member_df['年齡'].isnull().sum())

# 2. 将生日转换为datetime类型
member_df['生日'] = pd.to_datetime(member_df['生日'], errors='coerce')
print("转换后的生日缺失值数量:", member_df['生日'].isnull().sum())

# 3. 检查数据是否有重复会员卡号
duplicate_cards = member_df['会员卡号'].duplicated().sum()
print(f"重复的会员卡号数量: {duplicate_cards}")

# 如果有重复卡号，查看一些示例
if duplicate_cards > 0:
    print("\n重复卡号示例:")
    duplicated_card_examples = member_df[member_df['会员卡号'].duplicated(keep=False)].sort_values('会员卡号').head(10)
    print(duplicated_card_examples[['会员卡号', '会员等级', '注册时间', '性别', '年齡']])

处理后的年龄缺失值数量: 17
转换后的生日缺失值数量: 168157
重复的会员卡号数量: 8104

重复卡号示例:
             会员卡号  会员等级       注册时间 性别    年齡
700742  000353AAC  白银会员 2017-08-03  女   0.0
803457  000353AAC  白银会员 2017-08-03  女   0.0
684956  000390A6D  白银会员 2017-08-03  女  36.0
819576  000390A6D  白银会员 2017-08-03  女  26.0
683824  000393A59  白银会员 2017-08-03  女  28.0
818711  000393A59  白银会员 2017-08-03  女   0.0
804210  000639B9A  白银会员 2017-08-03  女   0.0
722183  000639B9A  白银会员 2017-08-03  女   0.0
716655  000649B39  白银会员 2017-08-03  女  34.0
736707  000649B39  白银会员 2017-08-03  女   0.0


8104条会员卡号重复，这里为了进行分析简单处理 去重取第一条

In [9]:
# 保留每个会员卡号的第一条记录
member_df_no_duplicates = member_df.drop_duplicates(subset=['会员卡号'], keep='first')

# 查看去重前后的数据量
print(f"去重前记录数: {member_df.shape[0]}")
print(f"去重后记录数: {member_df_no_duplicates.shape[0]}")
print(f"移除的重复记录数: {member_df.shape[0] - member_df_no_duplicates.shape[0]}")

# 使用去重后的数据继续操作
member_df = member_df_no_duplicates

去重前记录数: 952714
去重后记录数: 944610
移除的重复记录数: 8104


去重完毕，看看会员等级和性别的值分布

In [10]:
# 1. 检查会员等级和性别是否存在异常值
member_df['会员等级'].value_counts()

会员等级
白银会员    800146
黄金会员    142874
铂金会员      1106
钻石会员       484
Name: count, dtype: int64

In [11]:
# 检查性别分布
member_df['性别'].value_counts()

性别
女     860582
男      84026
未知         2
Name: count, dtype: int64

In [12]:
# 检查生命级别分布
member_df['生命级别'].value_counts()

生命级别
沉睡    795419
活跃     63967
休眠     62408
互动     22816
Name: count, dtype: int64

创建缺失率函数，然后算出每列的缺失率

In [13]:
# 创建一个计算缺失率的函数
def missing_rate(df):
    """计算每列的缺失率"""
    missing = df.isnull().sum()
    missing_percent = (missing / len(df)) * 100
    missing_stats = pd.DataFrame({'缺失值数量': missing, '缺失率(%)': missing_percent})
    return missing_stats.sort_values('缺失率(%)', ascending=False)

# 计算并显示各列缺失率
missing_rate(member_df)

缺失值数量    缺失率(%)
生日      166041  17.57773
年齡          17   0.00180
会员卡号         0   0.00000
会员等级         0   0.00000
会员来源         0   0.00000
所属店铺编码       0   0.00000
注册时间         0   0.00000
性别           0   0.00000
生命级别         0   0.00000

In [14]:
# 删除生日缺失的行
rows_before = len(member_df)
member_df = member_df.dropna(subset=['生日'])
rows_after = len(member_df)

# 显示删除的行数和比例
f"删除前行数: {rows_before}，删除后行数: {rows_after}，删除了 {rows_before - rows_after} 行 ({((rows_before - rows_after) / rows_before) * 100:.2f}%)"

'删除前行数: 944610，删除后行数: 778569，删除了 166041 行 (17.58%)'

导出文件

In [ ]:
# 导出清洗后的会员信息数据
member_df.to_excel('会员信息查询_清洗后.xlsx', index=False)

# 显示导出的数据形状
f"已导出清洗后的数据，共 {member_df.shape[0]} 行 {member_df.shape[1]} 列"